### DISCUSSION TITLE
Preview feature feedback: `omitApi`, exclude fields from Prisma Client query results

### DISCUSSION THREAD

**Discussion by janpio** [prisma/prisma#23924] [Feedback]
Support for excluding fields from Prisma Client query results via `omit` is in preview behind the `omitApi` preview feature since release of 5.13.0 on April 23rd:

- Documentation: https://www.prisma.io/docs/orm/reference/prisma-client-reference#omit-preview

Please share your feedback about this new functionality in this discussion.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this discussion.
- If you have any questions, don't hesitate to ask them in the [Prisma Discord](https://pris.ly/discord) community

**Comment by Cat7373** [comment]
Hello, I encountered some problems. It seems that Result does not support the new omit feature for the time being:

https://github.com/prisma/prisma/discussions/23947

**Reply by janpio** [reply]
This could be reproduced and was turned into a bug report issue: https://github.com/prisma/prisma/issues/23990

**Comment by khalilsarwari** [comment]
The typings when using `omit` do not reflect what columns are being omitted (or maybe its just a local issue for me).

If I do
```
const prisma = new PrismaClient({
  omit: {
    user: { 
	    // make sure that password and internalId are never queried.
	    password: true,
	    internalId: true,
	  },
  },
});

const myuser = await prisma.user.findFirst({});
```

Typescript still autocompletes `myuser.internalId`, whereas if i didnt select this column, it would not show it as a property and underline in red. This would be helpful to make sure adding `omit` at the client level does not break our code.

Also can it be possible to add omit via a client extension? I dont want to add for all my clients, which is the only way currently unless i setup a new PrismaClient with its own connection.

**Reply by jdhenry08** [reply]
I've also noticed this. It makes the omitApi feature nearly pointless to use, since typesafety is a big part of Prisma's draw for our codebase.

**Reply by SevInf** [reply]
@khalilsarwari @jdhenry08 please open an issue and provide a reproduction. I was not able to reproduce this issue:

![image](https://github.com/prisma/prisma/assets/603056/995f15a0-a9e5-4e69-8862-e188cf116873)

(`password` field exists on the model)

**Reply by khalilsarwari** [reply]
@SevInf I am unable to reproduce the issue; I think it might have just been a bad type cast actually that I had done from this side for `PrismaClient`.

**Reply by khalilsarwari** [reply]
@SevInf I was able to reproduce the original issue: https://github.com/prisma/prisma/issues/24717
I also faced another issue with `omitApi` and `createMany`: https://github.com/prisma/prisma/issues/24716

Your help would be greatly appreciated!

**Reply by SevInf** [reply]
Thank you @khalilsarwari.
For original issue, I provided the solution (using `as const` on config variable).
And I can confirm `createMany` problem, will work on a fix soon.

**Comment by achego** [comment]
After adding the omitApi preview feature, i still cannot use it globally, it throws a typescript warning however, it is enabled locally, what could the issue be

<img width="376" alt="Screenshot 2024-07-07 at 14 55 24" src="https://github.com/prisma/prisma/assets/41843978/39cb6263-8a3f-42c1-bfb3-d06cacc90e6e">

**Reply by SevInf** [reply]
Couple of things to check:
1. Are you on the latest prisma version? Global omit is available only starting from 5.16.0, while local one can be used starting with 5.13.0
2. Did you re-ran `prisma generate` after updating?
3. Have you tried restarting VSCode after update?

**Reply by achego** [reply]
Yes, @SevInf 
1.  My prisma version  "prisma": "^5.15.0"
2. yes i did
3. yes ive done that multiple times

**Reply by SevInf** [reply]
@achego latest prisma version is `5.16.1` and global omit is not available in `5.15.0`

**Reply by achego** [reply]
oh, ok.. thank you very much

**Comment by juanmdpc** [comment]
There is an issue when I'm trying to use the latest available version of the Prisma plugin inside the newest version of Webstorm IDE
![image](https://github.com/user-attachments/assets/39a690a8-b87d-4fa2-9c00-3eee616c111f)

![image](https://github.com/user-attachments/assets/d6e7417f-fec3-4607-83a4-91bfc69a5907)


![image](https://github.com/user-attachments/assets/b2e0b54a-dffa-4136-aa42-67d6d506a6db)

**Comment by jacob-hutchings** [comment]
I might have found a bug with the global omit feature.

Specifying a single model generates the correct types with the omitted field gone:
```
new PrismaClient({
    omit: {
        firstModel: {
            secret: true,
        },
    },
});
```

Specifying multiple models generates incorrect types with all fields included:
```
new PrismaClient({
    omit: {
        firstModel: {
            secret: true,
        },
        secondModel: {
            secret: true,
        },
    },
});
```


```
"prisma": "^5.17.0",
"@prisma/client": "^5.17.0",
```

**Reply by AdarshHatkar** [reply]
me facing same issue

**Reply by demetrius-mp** [reply]
facing the same issue with prisma and prisma client version ^5.17.0

**Reply by muchisx** [reply]
@SevInf @janpio Same here, everything working fine until you add more than one field

**Reply by pastine** [reply]
I'm also facing this same issue with prisma & prisma client versions as `5.19.1`.

**Reply by jacob-hutchings** [reply]
I was able to solve this issue by declaring the config like this:
```
const omitConfig = {
    firstModel: {
        secret: true,
    },
    secondModel: {
        secret: true,
    },
} as const;

new PrismaClient({
    omit: omitConfig,
});
```

**Reply by muchisx** [reply]
That works @jacob-hutchings , thank you !

**Reply by crisshaker** [reply]
For anybody still facing this issue, a type-safe way to get around it:
```js
import { PrismaClient, Prisma } from '@prisma/client';

const omit = {
  collection: { privateKey: true, public: true },
  user: { password: true, email: true },
} satisfies Prisma.GlobalOmitConfig;

export const prisma = new PrismaClient({
  omit,
});
```

**Comment by bchurlinov** [comment]
I am using "prisma":"5.19.1 and I can confirm that doesn't work at all. Is there any specific version that this works ?

**Comment by giftlion** [comment]
We are using the new omitApi in our project to global omit the password from the user.
We also use the generated User type from the Prisma type exports.
That causes issues because the type is generated before creating the client, so it still includes the password field.
Is there any way to get the type of the model after omitting, i.e. as if calling the findFirst method of the user, without resorting to workarounds like ReturnType?

THX!

**Comment by fbialek** [comment]
omitApi doesn't work for me...
I tried version 5.17 and 5.19.
I have added:
```generator client {
  provider = "prisma-client-js"
  previewFeatures = ["omitApi"]
}
```
```
const omit = {
  user: {
    passwordHash: true,
  },
}
if (process.env.NODE_ENV === "production") {
  db = new PrismaClient({
    omit ,
  })
} else {
  if (!global.__db) {
    global.__db = new PrismaClient({
      omit,
    })
  }
  db = global.__db
}

export { db }
```

I run `npx prisma generate` but I still can see passwordHash being returned by my queries...

**Reply by khalilsarwari** [reply]
Have you tried the "as const"?
```
const omit = {
  user: {
    passwordHash: true,
  },
} as const
```

**Reply by fbialek** [reply]
Yes, I just tried. It still doesn't work...
The only effect after the update to 5.17 (previously I was on v4) is that `new Prisma()` accepts object with omit field and TS nicely suggest names of my models.... but I still can see omitted field in objects that are returned by my queries

**Comment by patjakubik** [comment]
Feature request: would be amazing if it would be possible to globally omit computed fields. We would like to do some expensive computing in custom fields that is rarely accessed in the app.

**Comment by tuanpq1998** [comment]
I am using 5.19.1 and it works, thank you

**Reply by gerkim62** [reply]
`"@prisma/client": "^5.19.1"`
    i have restarted vs code multiple times and added `previewFeatures = ["omitApi"]` but still doesn't work
    
    it says "Object literal may only specify known properties, and 'omit' does not exist in type 'Subset<PrismaClientOptions, PrismaClientOptions>'.ts(2353)"

![image](https://github.com/user-attachments/assets/3c9ad18e-429b-467e-b145-3a5eb99b98c8)
see the red squiggly above

**Reply by gerkim62** [reply]
i was able to fix it by just rerunning `npx prisma generate`

**Comment by dragos-popescu** [comment]
I have a problem with global omit in my Next.js project. Let's say I omit the password from the user model like this:
```typescript
const prisma = new PrismaClient({
  omit: {
    user: { 
	 password: true,
     },
  },
});

``` 
Then I have a route /user/update/[id] and in the server component page.tsx i get the user and send it to a Form component:
```typescript
const Page = async ({ params }: Props) => {
  const { id } = await params;
  const user = await prisma.user.findUnique({
    where: {
       id: id
        }. 
    });

   return <Form user={user} />
}

``` 
So the Form component is something like this:
```typescript
import { type User } from "@prisma/client";

export default function Form({ user }: {user: User}) {
   ....
}

``` 
The problem is that the type User from @prisma/client still has the password field so there is a type mismatch:
- In the Page component, the user's type is correctly inferred to not have the globally omitted password field;
- In the Form component, the imported user type from @prisma/client has the password field.
This results in a TypeScript error.

I would expect the type User from @prisma/client to be updated so as to not include the password.

**Reply by gerkim62** [reply]
Use Omit<User, "password"> as the Type for the user prop

**Reply by vniehues** [reply]
yes, additionally omitting it from the Type works but it goes against using prisma as the single source of truth for your database models/types. 
imo this is an important part of using prisma in the first place - I would like to see this fixed

**Reply by aqrln** [reply]
> I would expect the type User from @prisma/client to be updated so as to not include the password.

That doesn't make a lot of sense to me. The `User` model reflects your schema, not your client configuration. It's also not technically possible because there's no association between the `User` type and the client: passing configuration to the client can't magically rewrite independent types.

If you want a field to be completely absent from the client API and from the `User` model type, you should be `@ignore`ing it in the schema. If you `omit` it globally in the client configuration, this implies that you still want to un-omit and access it in specific queries — you just want to omit it in most queries by default to reduce the boilerplate and avoid mistakes, but the data model and types are still the same.

I understand the problem here, but I think fundamentally it's not a question about `omit`. If you want to use model types for the results of queries, those will contain extra fields in most cases, whether you use `omit` or not, unless you are overfetching or have a simple application with trivial data model. Most queries in real world applications only select a subset of fields, so you end up defining DTOs using `Pick`/`Omit` helpers or using something like `Awaited<ReturnType<myRepositoryMethod>>` anyway.
